runners/direct-java/src/main/java/org/apache/beam/runners/direct/WriteWithShardingFactory.java

```diff
@@ -60,9 +60,10 @@
   public PTransformReplacement<PCollection<InputT>, PDone> getReplacementTransform(
       AppliedPTransform<PCollection<InputT>, PDone, PTransform<PCollection<InputT>, PDone>>
           transform) {
-
     try {
-      WriteFiles<InputT> replacement = WriteFiles.to(WriteFilesTranslation.getSink(transform));
+      WriteFiles<InputT, ?, ?> replacement = WriteFiles.to(
+          WriteFilesTranslation.getSink(transform),
```

Comment:
The intent of use of XYZTranslation here is that the DirectRunner (hence any runner) is able to interact with either of (a) a pipeline/transform that was just constructed in Java or (b) a pipeline/transform that was deserialized from proto and may or may not even be representable as a particular Java class beyond "RehydratedPTransform".

You may view the XYZTranslation as another "Runner API" whereby a runner never - or only at the very last moment - interacts with SDK-specific UDFs and transform classes.